add tests for Promise.withResolvers proposal

peetklecha opened the pull request:
This shouldn't be merged yet as Promise.withResolvers is still only Stage 2.

ljharb commented on test/built-ins/Promise/withResolvers/result.js:
```diff
@@ -0,0 +1,18 @@
+/*---
+description: Promise.withResolvers result is an object with keys "promise", "reject", and "resolve"
+esid: sec-promise.withresolvers
+features: [promise-with-resolvers]
```
```suggestion
features: [promise-with-resolvers, Object.hasOwn]
```
unless you want to use `.hasOwnProperty` below

peetklecha replied:
thanks for this!

peetklecha commented:
Now Stage 3.

ljharb commented on test/built-ins/Promise/withResolvers/result.js:
```diff
@@ -0,0 +1,18 @@
+var instance = Promise.withResolvers();
+
+assert.sameValue(typeof instance, "object");
+assert.notSameValue(instance, null);
```
```suggestion
assert.notSameValue(instance, null);
assert(instance intanceof Object);
```

trflynn89 commented on features.txt:
```diff
@@ -108,6 +108,10 @@ json-parse-with-source
 iterator-helpers
 
+# Promise.withResolvers
+# https://github.com/tc30/proposal-promise-with-resolvers
```
Noticed a typo here - the link has `tc30` instead of `tc39`

ljharb replied:
```suggestion
# https://github.com/tc39/proposal-promise-with-resolvers
```

trflynn89 commented on test/built-ins/Promise/withResolvers/result.js:
```diff
@@ -0,0 +1,19 @@
+
+assert.sameValue(typeof instance, "object");
+assert.notSameValue(instance, null);
+assert(instance intanceof Object);
```
Typo here as well: `intanceof` should be `instanceof`.

ljharb replied:
whoops
```suggestion
assert(instance instanceof Object);
```

Ms2ger commented on test/built-ins/Promise/withResolvers/ctx-non-ctor.js:
```diff
@@ -0,0 +1,12 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+description: Promise.withResolvers errors when the reciever is not a constructor
```
```suggestion
description: Promise.withResolvers errors when the receiver is not a constructor
```

linusg commented on test/built-ins/Promise/withResolvers/ctx-ctor.js:
```diff
@@ -0,0 +1,16 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+description: Promise.withResolvers produces instances of the reciever
```
```suggestion
description: Promise.withResolvers produces instances of the receiver
```

linusg commented on test/built-ins/Promise/withResolvers/ctx-non-object.js:
```diff
@@ -0,0 +1,32 @@
+  Promise.withResolvers.call(86);
+});
+  
+  assert.throws(TypeError, function() {
+    Promise.withResolvers.call('string');
+  });
+  
+  assert.throws(TypeError, function() {
+    Promise.withResolvers.call(true);
+  });
+  
+  assert.throws(TypeError, function() {
+    Promise.withResolvers.call(Symbol());
+  });
```
indentation is a little off here :)

linusg commented on test/built-ins/Promise/withResolvers/result.js:
```diff
@@ -0,0 +1,19 @@
+assert(instance intanceof Object);
+assert(Object.hasOwn(instance, "promise"));
+assert(Object.hasOwn(instance, "resolve"));
+assert(Object.hasOwn(instance, "reject"));
```
Would be nice to also test the property attributes here, using `verifyProperty()` from `propertyHelper.js`:

https://github.com/tc39/test262/blob/9d7eb8018e3b74ae9bcb12fbe8db0dfce4ef6adf/test/built-ins/BigInt/length.js#L27-L32

linusg replied:
(you can omit `value` as it's already tested elsewhere - https://github.com/tc39/test262/blob/9d7eb8018e3b74ae9bcb12fbe8db0dfce4ef6adf/harness/propertyHelper.js#L67-L71)

ptomato commented:
Thanks for contributing this! Would you mind opening a tracking issue? An example is #3756. You don't have to write the whole testing plan, someone else can help with that, but we should use one to verify that we're covering everything in the proposal.

I haven't looked too closely at this PR yet so I might be missing something, but it seems like there are some cases particularly with subclassing `Promise` that aren't yet tested.

We also need to add the standard boilerplate tests for built-in APIs. For an example, see `length.js`, `name.js`, `prop-desc.js`, `not-a-constructor.js`, `builtin.js` in https://github.com/tc39/test262/tree/main/test/built-ins/Array/fromAsync.

We don't have to block reviewing and merging this PR on complete coverage, the other tests can come in following PRs, but we should keep track in the tracking issue of what's been merged.

syg commented:
Could we have this merged soon for shipping in Chrome?

peetklecha commented:
sorry for the delay with this.

- i've now addressed all the comments i've recieved
- there's a failing test from circle/ci which wasn't there before and which i don't understand, any help with it would be appreciated
- per @ptomato 's request i've created an issue to track full test coverage of Promise.withResolvers: #3910

ptomato commented:
Thanks, @peetklecha. The failing CI job is #3891, we really should remove the job until that's fixed. I'll merge this now, and add some suggestions to the tracking issue you opened.
